oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/pipelined/PipelinedMongoDownloadTask.java

```diff
@@ -353,63 +354,101 @@ private void downloadWithNaturalOrdering() throws InterruptedException, TimeoutE
+        Set<String> excludedPaths = new HashSet<>();
+        for (PathFilter pathFilter : pathFilters) {
+            includedPaths.addAll(pathFilter.getIncludedPaths());
+            excludedPaths.addAll(pathFilter.getExcludedPaths());
+        }
+        // Sort by length to make it easier to compute the common ancestors of include paths
+        List<String> sortedIncludedPaths = includedPaths.stream()
+                .sorted()
+                .collect(Collectors.toList());
+
+        LOG.info("Paths considered for regex filtering. IncludedPaths: {}, excludedPaths: {}", sortedIncludedPaths, excludedPaths);
+        if (sortedIncludedPaths.contains("/")) {
+            LOG.info("Disabling regex path filtering because root path is in the includedPaths: {}", sortedIncludedPaths);
+            return Set.of();
+        }
+        if (!excludedPaths.isEmpty()) {
+            LOG.info("Disabling regex path filtering because there are excluded paths: {}", excludedPaths);
+            return Set.of();
         }
-        Set<String> includedPaths = pathFilters.stream()
-                .flatMap(pathFilter -> pathFilter.getIncludedPaths().stream())
-                .collect(Collectors.toSet());
-
-        Set<String> excludedPaths = pathFilters.stream()
-                .flatMap(pathFilter -> pathFilter.getExcludedPaths().stream())
-                .collect(Collectors.toSet());
 
-        if (excludedPaths.isEmpty() && includedPaths.size() == 1) {
-            return includedPaths.stream().iterator().next();
-        } else {
-            return null;
+        // Keep only unique include paths. That is, if paths "/a/b" and "/a/b/c" are both in the list, keep only "/a/b"
+        HashSet<String> includedPathsRoots = new HashSet<>();
+        for (String path : sortedIncludedPaths) {
+            if (includedPathsRoots.stream().noneMatch(ancestor -> PathUtils.isAncestor(ancestor, path))) {
+                includedPathsRoots.add(path);
+            }
         }
+        return includedPathsRoots;
     }
 
-    private static Bson descendantsFilter(String path) {
+    private static Bson descendantsFilter(Set<String> path) {
+        List<Bson> filters = path.stream()
+                .flatMap(PipelinedMongoDownloadTask::descendantsFilter)
+                .collect(Collectors.toList());
+        return Filters.or(filters);
+    }
+
+    private static Stream<Bson> descendantsFilter(String path) {
         if (!path.endsWith("/")) {
             path = path + "/";
         }
         String quotedPath = Pattern.quote(path);
         // For entries with path sizes above a certain threshold, the _id field contains a hash instead of the path of
         // the entry. The path is stored instead in the _path field. Therefore, we have to include in the filter also
         // the documents with matching _path.
-        return Filters.or(
+        return Stream.of(
                 regex(NodeDocument.ID, Pattern.compile("^[0-9]{1,3}:" + quotedPath + ".*$")),
```

Comment:
Is this condition correct.  The getIdFromPath has a different format.
https://github.com/apache/jackrabbit-oak/blob/0b8f4ab2e736c6561ae745a5fe6040a59911eeb3/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java#L291C61-L291C61

Reply:
Yeah this condition looks fine. 
Another query:
As we won't be able to filter all documents based on condition1, mongo would have to anyways traverse through all documents, would keeping condition1 make any difference?

Reply:
We need both conditions. The `_path` field is only present on documents with a "long path", that is, where the `_id` field was replaced by an hash. For instance:

https://github.com/apache/jackrabbit-oak/blob/0b8f4ab2e736c6561ae745a5fe6040a59911eeb3/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeState.java#L446-L459